### PR TITLE
K8SPG-911: Add pg_tde support

### PR DESCRIPTION
**Pull request by egegunes**
**CHANGE DESCRIPTION**
---
This commit adds native pg_tde extension support into operator.

**This commit only adds Vault KMS support for pg_tde. KMIP support will be added in future releases.**

When pg_tde is enabled and Vault configuration is provided, the operator:
- appends pg_tde into shared_preload_libraries,
- mounts Vault token and CA secrets into database containers,
- runs CREATE EXTENSION in all databases,
- creates Vault provider by running pg_tde_add_global_key_provider_vault_v2,
- create a global key by running pg_tde_create_key_using_global_key_provider,
- sets the default key by running pg_tde_set_default_key_using_global_key_provider.

## Example configuration

```
pg_tde:
  enabled: true
  vault:
    host: https://vault-service.vault-service.svc:8200
    mountPath: tde
    tokenSecret:
      name: vault-secret
      key: token
    caSecret:
      name: vault-secret
      key: ca.crt
```

Note that:
- Mount path needs to be a KV v2 storage engine. 
- caSecret is optional and can be omitted if you want to use http. But in my testing I couldn't manage the make vault work without TLS. It responds with HTTP 405 if I disable TLS in vault.
- tokenSecret and caSecret can be the same secret or different. Operator doesn't assume anything about the contents of the secrets since you'll need to set secret keys in cr.yaml yourself.
- Using a non-root token requires more configuration. Check out pg_tde docs for that. But don't forget to add these in the Vault policy:

```
path "sys/internal/ui/mounts/*" {
      capabilities = ["read"]
}
    
path "sys/mounts/*" {
      capabilities = ["read"]
}
```

## API changes

pg_tde requires more configuration options than other extensions operator supports. This required us make some changes in the extensions API. With these changes, 'spec.extensions.builtin' section is deprecated and all builtin extensions are moved to 'spec.extensions.<extension>' (i.e. 'spec.extensions.pg_stat_monitor'). Right now extensions can be enabled/disabled with the old and the new method. If two methods are used at the same time, 'spec.extensions.builtin' takes precedence.

## Status changes

A hash will be calculated using pg_tde configuration provided by user. Operator uses this hash to understand if config is changed and it should reconfigure pg_tde. The hash can be found in status.pgTDERevision field of **PostgresCluster** object. This hash will be removed when pg_tde is disabled.

Operator also communicates the status of pg_tde with conditions. The condition with type=PGTDEEnabled can be found in both PerconaPGCluster and PostgresCluster statuses.

## Disabling pg_tde

Disabling pg_tde is more complex than other extensions:
- First of all any encrypted objects must be dropped before disabling. Otherwise DROP EXTENSION will fail with a descriptive error message. **Operator won't drop anything, user needs to do this manually.**
- The extension needs to be disabled in two steps:
  1. First set pg_tde.enabled=false without removing the vault section. Operator will drop the extension and restart the pods.
  2. Then you can remove pg_tde.vault. Database pods will be restarted again to remove secret mounts from containers.
- It's recommended to run CHECKPOINT before removing pg_tde.vault. Even though extension is dropped, Postgres might still try to use encrypted objects during recovery after restart and it might try to access token secret. CHECKPOINT helps you prevent this failure case.

## Deleting and recreating clusters

If cluster with pg_tde enabled is deleted but PVCs are retained, on recreation you'll see some errors about pg_tde in operator logs. They happen because the vault provider and/or global key already exists. Operator will handle these errors gracefully and configure pg_tde. Same thing applies when pg_tde is disabled and re-enabled. Since both vault provider and global key already exists, operator will handle "already exists" errors and configure pg_tde.

The global key name is determined by cluster's .metadata.uid. For example 'global-master-key-ad19534a-d778-460e-ac87-ca38ef5e6755'. This means the key will be changed if cluster is deleted and recreated. As long as the old key and the new key is accessible to pg_tde, this won't cause any issues. pg_tde will handle it as it handles key rotation.

## Validations

- You can't set pg_tde.enabled=true without setting pg_tde.vault.
- If you already had pg_tde.enabled, you can't remove pg_tde section completely.
- If you already had pg_tde.enabled, you can't remove pg_tde.vault section completely.

## Multi platform testing

- [ ] EKS
- [x] AKS
- [x] Openshift

### CODE REVIEW

**Comment by matserix** [comment]
This looks really interesting, definitely would help to reduce doing some workarounds.

As KMIP is the standard protocol which is also supported by Vault and most cloud providers integrate it into their secret manager, it should/must be supported besides Vault.

**Comment by egegunes** [comment]
> This looks really interesting, definitely would help to reduce doing some workarounds.
> 
> As KMIP is the standard protocol which is also supported by Vault and most cloud providers integrate it into their secret manager, it should/must be supported besides Vault.

@matserix we decided to just start with Vault. KMIP support will be added in future releases.

**Comment by JNKPercona** [comment]
| Test Name | Result | Time |
| ----------- | -------- | ------ |
| backup-enable-disable | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/backup-enable-disable.log) | 00:07:18 |
| builtin-extensions | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/builtin-extensions.log) | 00:01:52 |
| cert-manager-tls | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/cert-manager-tls.log) | 00:04:50 |
| custom-envs | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/custom-envs.log) | 00:18:41 |
| custom-extensions | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/custom-extensions.log) | 00:01:52 |
| custom-tls | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/custom-tls.log) | 00:07:14 |
| database-init-sql | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/database-init-sql.log) | 00:02:06 |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/demand-backup.log) | 00:22:51 |
| demand-backup-offline-snapshot | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/demand-backup-offline-snapshot.log) | 00:13:29 |
| dynamic-configuration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/dynamic-configuration.log) | 00:03:01 |
| finalizers | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/finalizers.log) | 00:03:43 |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/init-deploy.log) | 00:02:47 |
| huge-pages | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/huge-pages.log) | 00:03:07 |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/monitoring.log) | 00:06:55 |
| monitoring-pmm3 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/monitoring-pmm3.log) | 00:08:08 |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/one-pod.log) | 00:05:47 |
| operator-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/operator-self-healing.log) | 00:10:12 |
| pg-tde | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/pg-tde.log) | 00:08:37 |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/pitr.log) | 00:11:44 |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/scaling.log) | 00:04:45 |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/scheduled-backup.log) | 00:24:35 |
| self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/self-healing.log) | 00:10:39 |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/sidecars.log) | 00:02:48 |
| standby-pgbackrest | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/standby-pgbackrest.log) | 00:11:45 |
| standby-streaming | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/standby-streaming.log) | 00:11:33 |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/start-from-backup.log) | 00:10:23 |
| tablespaces | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/tablespaces.log) | 00:06:39 |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/telemetry-transfer.log) | 00:04:31 |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/upgrade-consistency.log) | 00:06:24 |
| upgrade-minor | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/upgrade-minor.log) | 00:05:18 |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/8e995a580/users.log) | 00:04:18 |

| Summary | Value |
| ------- | ----- |
| Tests Run | 31/31 |
| Job Duration | 01:24:44 |
| Total Test Time | 04:08:09 |

commit: https://github.com/percona/percona-postgresql-operator/pull/1440/commits/8e995a580124520be977935d6b372eb001835ee1
image: `perconalab/percona-postgresql-operator:PR-1440-8e995a580`

File: pkg/apis/pgv2.percona.com/v2/perconapgcluster_types.go [hunk 1]
```diff
@@ -258,24 +257,62 @@ func (cr *PerconaPGCluster) Default() {
 		}
 	}
 
+	cr.SetExtensionDefaults()
+
+	if cr.CompareVersion("2.6.0") >= 0 && cr.Spec.AutoCreateUserSchema == nil {
```
Comment (nmarukovich) [inline pkg/apis/pgv2.percona.com/v2/perconapgcluster_types.go:262]: I think we can already delete this condition: `cr.CompareVersion("2.6.0")`

File: e2e-tests/tests/pg-tde/08-disable-pgtde.yaml
Comment (eleo007) [inline e2e-tests/tests/pg-tde/08-disable-pgtde.yaml:1]: Can we add to test check that after disabling the data is written and has disabled encryption?

File: e2e-tests/tests/pg-tde/06-assert.yaml
Comment (eleo007) [inline e2e-tests/tests/pg-tde/06-assert.yaml:1]: Can we add check to verify restored data?

**Review by eleo007 — COMMENTED** [review]
Use new extensions structure in e2e-tests/tests/builtin-extensions/03-install-all-ext.yaml

File: internal/controller/postgrescluster/postgres.go [hunk 2]
```diff
@@ -336,6 +341,19 @@ func (r *Reconciler) reconcilePostgresDatabases(
 			}
 		}
 
+		// K8SPG-911
```
Comment (gkech) [inline internal/controller/postgrescluster/postgres.go:343]: If possible, let's extract all this pgtde logic from the `reconcilePostgresDatabases` function. We are essentially doing 2 things here that are scattered around the function. 

1) run the enable/disable flow
2) some lines later we set the status for this feature through `SetStatusCondition`

We could have a function like `func (r *Reconciler) reconcilePGTDE(
	ctx context.Context,
	cluster *v1beta1.PostgresCluster,
	exec postgres.Executor,
) bool `

that handles all this pgtde logic in a single place, calling it through on this like

```
		// K8SPG-911
		pgTdeOK = r.reconcilePGTDE(ctx, cluster, exec)
```

File: internal/controller/postgrescluster/postgres.go [hunk 3]
```diff
@@ -375,14 +393,122 @@ func (r *Reconciler) reconcilePostgresDatabases(
...
+}
+
+// reconcilePGTDEProviders configures pg_tde providers
+func (r *Reconciler) reconcilePGTDEProviders(
+	ctx context.Context,
+	cluster *v1beta1.PostgresCluster,
+	instances *observedInstances,
+	patchStatus func() error,
+) error {
+	var podExecutor postgres.Executor
```
Comment (gkech) [inline internal/controller/postgrescluster/postgres.go]: Not sure why we need to define this var here. Couldn't we just `	podExecutor := func(
		ctx context.Context, stdin io.Reader, stdout, stderr io.Writer, command ...string,
	) error {
		return r.PodExec(ctx, pod.Namespace, pod.Name, container, stdin, stdout, stderr, command...)
	}` some lines later?

File: internal/controller/postgrescluster/postgres.go [hunk 4]
```diff
@@ -375,14 +393,122 @@ func (r *Reconciler) reconcilePostgresDatabases(
...
+	}
+
+	// We need to configure pg_tde after volumes are mounted and extension is created
+	if _, ok := pod.Annotations[naming.TDEInstalledAnnotation]; !ok {
+		return nil
+	}
+
+	log := logging.FromContext(ctx).WithName("PGTDE")
+
+	ctx = logging.NewContext(ctx, logging.FromContext(ctx).WithValues("pod", pod.Name))
```
Comment (gkech) [inline internal/controller/postgrescluster/postgres.go]: Maybe something more clear since  fields are split across log and ctx

```
 log := logging.FromContext(ctx).WithName("PGTDE").WithValues("pod", pod.Name)
  ctx = logging.NewContext(ctx, log)
```

?

File: internal/controller/postgrescluster/postgres.go [hunk 5]
```diff
@@ -375,14 +393,122 @@ func (r *Reconciler) reconcilePostgresDatabases(
...
+		}
+		err = errors.WithStack(err)
+	}
+
+	if err == nil {
+		cluster.Status.PGTDERevision = revision
+		if err := patchStatus(); err != nil {
+			return errors.Wrap(err, "patch status")
+		}
 	}
```
Comment (gkech) [inline internal/controller/postgrescluster/postgres.go]: here we have 2 blocks with the same if condition? Can we simplify?

File: internal/controller/postgrescluster/postgres.go [hunk 6]
```diff
@@ -375,14 +393,122 @@ func (r *Reconciler) reconcilePostgresDatabases(
...
+			if err == nil || errors.Is(err, pgtde.ErrAlreadyExists) {
+				err = pgtde.CreateGlobalKey(ctx, podExecutor, cluster.UID)
+			}
+			if err == nil || errors.Is(err, pgtde.ErrAlreadyExists) {
+				err = pgtde.SetDefaultKey(ctx, podExecutor, cluster.UID)
+			}
+		} else {
+			err = pgtde.ChangeVaultProvider(ctx, podExecutor, vault)
+		}
+		err = errors.WithStack(err)
```
Comment (gkech) [inline internal/controller/postgrescluster/postgres.go]: Recommend we extract this to a function, we set the `err` in 4 different places and then we overwrite it with `errors.WithStack(err)` to handle it afterwards with the next if clause. It will be easier to reason.

File: internal/pgtde/postgres_test.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+		assert.Equal(t, string(b), strings.Join([]string{
+			`SET client_min_messages = WARNING;`,
+			`CREATE EXTENSION IF NOT EXISTS pg_tde;`,
+			`ALTER EXTENSION pg_tde UPDATE;`,
+		}, "\n"))
+
+		return expected
+	}
+
+	ctx := context.Background()
```
Comment (gkech) [inline internal/pgtde/postgres_test.go]: we can use the ctx of the test `t.Context()`

File: internal/pgtde/postgres_test.go [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+				Name: "token-secret",
+				Key:  "token-key",
+			},
+		}
+		assert.NilError(t, AddVaultProvider(ctx, exec, vault))
+	})
+}
+
+func TestCreateGlobalKey(t *testing.T) {
+	t.Run("success", func(t *testing.T) {
```
Comment (gkech) [inline internal/pgtde/postgres_test.go:184]: TestCreateGlobalKey and TestSetDefaultKey "success" subtests are misnamed most probably since both only verify the whoops case. Is that correct?

Comment (gkech) [reply internal/pgtde/postgres_test.go:184]: should we also test the nil error case in general?

File: internal/controller/postgrescluster/controller.go [hunk 9]
```diff
@@ -273,6 +274,15 @@ func (r *Reconciler) Reconcile(
...
 	}
+
+	pgTDECondition := meta.FindStatusCondition(cluster.Status.Conditions,
+		v1beta1.PGTDEEnabled)
+	pgTDEEnabled := pgTDECondition != nil && pgTDECondition.Status == metav1.ConditionTrue
+	// pg_tde should be removed from shared libraries only after extension is dropped
+	if cluster.Spec.Extensions.PGTDE.Enabled || pgTDEEnabled {
+		pgtde.PostgreSQLParameters(&pgParameters)
+	}
+
```
Comment (pooknull) [inline internal/controller/postgrescluster/controller.go:285]: What if the cluster is recreated with existing PVCs where the `pg_tde` extension had previously been enabled? As I understand it, the new cluster would not have the status condition set, causing this check to be false. As a result, `pg_tde` would be removed from shared libraries without the extension being dropped first.

File: internal/controller/postgrescluster/postgres.go (hunk 2 above)
Comment (egegunes) [reply internal/controller/postgrescluster/postgres.go:343]: enabling extension needs to be done in `reconcilePostgresDatabases` because it's affected by `DatabaseRevision`. `reconcilePGTDEProviders` only configures pg_tde. I think it should stay like this.

File: internal/pgtde/postgres_test.go (hunk 8 above)
Comment (egegunes) [reply internal/pgtde/postgres_test.go:184]: I don't know why but this pattern was used in upstream and we use the same in every unit tests for extensions.

I don't think testing the nil case will make any difference. Do you have a case in mind?

File: internal/controller/postgrescluster/controller.go (hunk 9 above)
Comment (egegunes) [reply internal/controller/postgrescluster/controller.go:285]: this condition checks two things: either `cluster.Spec.Extensions.PGTDE.Enabled` is true **or** condition is true.

if cluster is deleted and recreated with pg_tde enabled, it'll be added into shared libraries.

File: pkg/apis/pgv2.percona.com/v2/perconapgcluster_types.go (hunk 1 above)
Comment (egegunes) [reply pkg/apis/pgv2.percona.com/v2/perconapgcluster_types.go:262]: is it okay if we remove it next release?

File: e2e-tests/tests/pg-tde/08-disable-pgtde.yaml
Comment (egegunes) [reply e2e-tests/tests/pg-tde/08-disable-pgtde.yaml:1]: we check if extension is disabled by https://github.com/percona/percona-postgresql-operator/blob/954c47cbe5b7be8f532d3a73ccd20e5abe890bba/e2e-tests/tests/pg-tde/08-assert.yaml#L79-L82

but I can't check the data because I need to drop the table to disable extension

File: internal/pgtde/postgres.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+		if err == nil || errors.Is(err, errAlreadyExists) {
+			err = createGlobalKey(ctx, exec, cluster.UID)
+		}
+
+		if err == nil || errors.Is(err, errAlreadyExists) {
+			err = setDefaultKey(ctx, exec, cluster.UID)
+		}
+
+		return err
```
[the remark this answers is not in the document]
Comment (egegunes) [reply internal/pgtde/postgres.go:251]: i don't get it. if createGlobalKey returns errAlreadyExists, it already exists.

**Review by gkech — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test Name | Result | Time |
| ----------- | -------- | ------ |
| backup-enable-disable | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/backup-enable-disable.log) | 00:07:12 |
| builtin-extensions | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/builtin-extensions.log) | 00:05:20 |
| cert-manager-tls | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/cert-manager-tls.log) | 00:04:57 |
| custom-envs | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/custom-envs.log) | 00:18:33 |
| custom-extensions | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/custom-extensions.log) | 00:16:00 |
| custom-tls | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/custom-tls.log) | 00:06:48 |
| database-init-sql | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/database-init-sql.log) | 00:02:34 |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/demand-backup.log) | 00:22:49 |
| demand-backup-offline-snapshot | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/demand-backup-offline-snapshot.log) | 00:13:19 |
| dynamic-configuration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/dynamic-configuration.log) | 00:03:25 |
| finalizers | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/finalizers.log) | 00:03:44 |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/init-deploy.log) | 00:02:48 |
| huge-pages | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/huge-pages.log) | 00:02:53 |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/monitoring.log) | 00:07:14 |
| monitoring-pmm3 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/monitoring-pmm3.log) | 00:08:04 |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/one-pod.log) | 00:05:45 |
| operator-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/operator-self-healing.log) | 00:10:18 |
| pg-tde | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/pg-tde.log) | 00:09:15 |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/pitr.log) | 00:12:02 |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/scaling.log) | 00:05:07 |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/scheduled-backup.log) | 00:27:52 |
| self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/self-healing.log) | 00:07:57 |
| sidecars | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/sidecars.log) | 00:02:42 |
| standby-pgbackrest | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/standby-pgbackrest.log) | 00:13:01 |
| standby-streaming | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/standby-streaming.log) | 00:09:20 |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/start-from-backup.log) | 00:12:18 |
| tablespaces | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/tablespaces.log) | 00:06:50 |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/telemetry-transfer.log) | 00:06:36 |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/upgrade-consistency.log) | 00:04:54 |
| upgrade-minor | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/upgrade-minor.log) | 00:05:57 |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-1440/fc2a47f54/users.log) | 00:04:31 |

| Summary | Value |
| ------- | ----- |
| Tests Run | 31/31 |
| Job Duration | 01:25:57 |
| Total Test Time | 04:30:20 |

commit: https://github.com/percona/percona-postgresql-operator/pull/1440/commits/fc2a47f54865e0951ec1534364852968e0de7cf3
image: `perconalab/percona-postgresql-operator:PR-1440-fc2a47f54`
